### PR TITLE
Add support for Tongue TOQCB2-80 

### PR DESCRIPTION
**Pull request by bifi2090**
Add support for Tongue TOQCB2-80 from

https://github.com/Koenkk/zigbee2mqtt/issues/21943#issuecomment-2212419031

### CODE REVIEW

**Comment by bifi2090** [comment]
OK Guys, i need some help. I don´t know, if i must put something like:

```
TOQCB2_settings: {
    key: ['temperature_setting','temperature_threshold','under_voltage_setting','under_voltage_threshold',
          'over_current_setting','current_threshold','over_voltage_setting','over_voltage_threshold',
          'over_power_setting','over_power_threshold','restore_default'],
    convertSet: async (entity, key, value, meta) => {
        const settingType = {'trip': 2, 'alarm': 1, 'closed': 0};
        // const typeswitch = {'ON':true, 'OFF':false}
        let dp_key = -1; // value if not found
        // identify dp.
        const tuyadatapoints_list = meta.mapped.meta.tuyaDatapoints
        for (let i = 0; i < tuyadatapoints_list.length; i++) {
            if (tuyadatapoints_list[i][1] === key) {
                dp_key = tuyadatapoints_list[i][0];
                break;
            }
        }
        switch (key) {
        case 'temperature_setting': {
            await tuya.sendDataPointEnum(entity, dp_key, utils.getFromLookup(value, settingType));
            break;
        }
        case 'temperature_threshold': {
            await tuya.sendDataPointValue(entity, dp_key, value*10);
            break;
        }
        case 'under_voltage_setting': {
            await tuya.sendDataPointEnum(entity, dp_key, utils.getFromLookup(value, settingType));
            break;
        }
        case 'under_voltage_threshold': {
            await tuya.sendDataPointValue(entity, dp_key, value);
            break;
        }
        case 'over_voltage_setting': {
            await tuya.sendDataPointEnum(entity, dp_key, utils.getFromLookup(value, settingType));
            break;
        }
        case 'over_voltage_threshold': {
            await tuya.sendDataPointValue(entity, dp_key, value);
            break;
        }
        case 'over_power_setting': {
            await tuya.sendDataPointEnum(entity, dp_key, utils.getFromLookup(value, settingType));
            break;
        }
        case 'over_power_threshold': {
            await tuya.sendDataPointValue(entity, dp_key, value);
            break;
        }
        case 'over_current_setting': {
            await tuya.sendDataPointEnum(entity, dp_key, utils.getFromLookup(value, settingType));
            break;
        }
        case 'current_threshold': {
            await tuya.sendDataPointValue(entity, dp_key, value);
            break;
        }
        default: // Unknown key
            meta.logger.warn(`Unhandled key ${key}`);
        }
    },
}

```

somewhere between for example line 441 and 442. It gives me a lot of errors at build time.

File: src/devices/tuya.ts [hunk 1]
```diff
@@ -10458,6 +10458,157 @@ const definitions: Definition[] = [
...
+            tuya.exposes.voltageWithPhase('a'), 
+            tuya.exposes.voltageWithPhase('b'), 
+            tuya.exposes.voltageWithPhase('c'),
+            tuya.exposes.powerWithPhase('a'), 
+            tuya.exposes.powerWithPhase('b'), 
+            tuya.exposes.powerWithPhase('c'),
+            tuya.exposes.currentWithPhase('a'), 
+            tuya.exposes.currentWithPhase('b'), 
+            tuya.exposes.currentWithPhase('c'),
+            e.numeric('forward_electricity', ea.STATE).withUnit('kwh'),
```
Comment (Koenkk) [inline src/devices/tuya.ts:10488]: Can you use `e.produced_energy()` for this?

File: src/devices/tuya.ts [hunk 2]
```diff
@@ -10458,6 +10458,157 @@ const definitions: Definition[] = [
...
+            e.enum('last_event1', ea.STATE,
```
Comment (Koenkk) [inline src/devices/tuya.ts:10489]:
```suggestion
            e.enum('last_event', ea.STATE, 
```

File: src/devices/tuya.ts [hunk 3]
```diff
@@ -10458,6 +10458,157 @@ const definitions: Definition[] = [
...
+            tuya.exposes.powerWithPhase('a'), 
+            tuya.exposes.powerWithPhase('b'), 
+            tuya.exposes.powerWithPhase('c'),
+            tuya.exposes.currentWithPhase('a'), 
+            tuya.exposes.currentWithPhase('b'), 
+            tuya.exposes.currentWithPhase('c'),
+            e.numeric('forward_electricity', ea.STATE).withUnit('kwh'),
+            e.enum('last_event1', ea.STATE, 
+                   ['Normal', 
+                    'Trip_Over-current',
```
Comment (Koenkk) [inline src/devices/tuya.ts:10491]:
```suggestion
                    'trip_over_current', 
```

Please make all the entries lower snake case.

File: src/devices/tuya.ts (hunk 2 above)
Comment (bifi2090) [reply src/devices/tuya.ts]: this works also

**Comment by Koenkk** [comment]
Thanks!

**Comment by Proncio** [comment]
Hi,
2 small hints:

1. Vendor is Tongou, rather than generic Tuya
2. Last event is not working for me, looking at the code DP name in Expose and Meta sections is different, last_event1 vs last_event, changing meta one to last_event it works

**Pull request merged**